core/src/banking_simulation.rs

```diff
@@ -0,0 +1,616 @@
+        event_file_path: &PathBuf,
+    ) -> Result<(), SimulateError> {
+        let mut reader = BufReader::new(File::open(event_file_path)?);
+
+        loop {
+            events.push(deserialize_from(&mut reader)?);
+
+            if reader.fill_buf()?.is_empty() {
+                // EOF is reached at a correct deserialization boundary.
+                // We're looking-ahead the buf, so NOT calling reader.consume(..) is correct.
+                break;
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn load(event_file_paths: Vec<PathBuf>) -> Result<Self, SimulateError> {
+        let mut events = vec![];
+        for event_file_path in &event_file_paths {
+            let old_len = events.len();
+            let _ = Self::read_event_file(&mut events, event_file_path).inspect_err(|error| {
+                error!(
+                    "Reading {event_file_path:?} failed after {} events: {:?} due to file corruption or unclearn validator shutdown",
+                    events.len() - old_len,
+                    error
+                );
+            });
+            info!(
+                "Read {} events from {:?}",
+                events.len() - old_len,
+                event_file_path
+            );
+        }
+
+        let mut packet_batches_by_time = BTreeMap::new();
+        let mut freeze_time_by_slot = BTreeMap::new();
+        let mut hash_overrides = HashOverrides::default();
+        for TimedTracedEvent(event_time, event) in events {
+            match event {
+                TracedEvent::PacketBatch(label, batch) => {
+                    // Somewhat naively assume that event_times (nanosecond resolution) won't
+                    // collide.
+                    let is_new = packet_batches_by_time
+                        .insert(event_time, (label, batch))
+                        .is_none();
+                    assert!(is_new);
+                }
+                TracedEvent::BlockAndBankHash(slot, blockhash, bank_hash) => {
+                    let is_new = freeze_time_by_slot.insert(slot, event_time).is_none();
+                    hash_overrides.add_override(slot, blockhash, bank_hash);
+                    assert!(is_new);
+                }
+            }
+        }
+
+        Ok(Self {
+            packet_batches_by_time,
+            freeze_time_by_slot,
+            hash_overrides,
+        })
+    }
+
+    pub fn hash_overrides(&self) -> &HashOverrides {
+        &self.hash_overrides
+    }
+}
+
+impl BankingSimulator {
+    pub fn new(banking_trace_events: BankingTraceEvents, first_simulated_slot: Slot) -> Self {
+        Self {
+            banking_trace_events,
+            first_simulated_slot,
+        }
+    }
+
+    pub fn parent_slot(&self) -> Option<Slot> {
+        self.banking_trace_events
+            .freeze_time_by_slot
+            .range(..self.first_simulated_slot)
+            .last()
+            .map(|(slot, _time)| slot)
+            .copied()
+    }
+
+    pub fn start(
```

Comment:
feel pretty strongly this function is too big and readability would greatly benefit from breaking it up.

It seems like it has several distinct phases and/or threads already:
1. setup
2. sender-thread
3. main-loop

Reply:
phew... finally done: a83587d471ce15736f6d3ffee7bf2d9229aa7013